---
title: "RFdecd Package Vignette"
package: RFdecd
output:
  BiocStyle::html_document:
    toc: true
    toc_float:
      collapsed: true
      smooth_scroll: true
vignette: >
  %\VignetteIndexEntry{RFdecd Package Vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
```{r knitr-options, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(RFdecd)
```

# Description of `RFdecd`
RFdecd is an R package for feature selection-based reference-free deconvolution.
It contains two main steps. Step (a) uses the reduced matrix obtained from the top 1000 features with the largest coefficient of variation on the raw data to perform a reference-free deconvolution to estimate the cell-type proportion matrix. Step (b) uses the cell-type proportion matrix estimated in step (a) and raw data to identify cell-type specific features by six feature selection options (VAR, CV, 1_O, 2_O, P_W and RFdecd). 


# Arguments of `RFdecd`
We can run `RFdecd` in the following code:
```{r}
res_RFdecd <- RFdeconv(DataType = "Gene expression",
                       Y_raw,
                       K,
                       CTSoption = DEVarSelect_RFdecd,
                       nMarker = 1000,
                       InitMarker = NULL,
                       TotalIter = 30)
```
In the function call above, 
`DataType` is covariate representing data type, either Gene expression or DNA methylation. 

`Y_raw` is a raw data matrix of complex samples. 

`K` is the number of cell types.  
`CTSoption` stands for six feature selection options:(1) DEVarSelect_CV for selecting the top 1000 features with the largest coefficient of variation (i.e., CV) in the estimated cell-type profiles; (2) DEVarSelect_VAR for selecting the top 1000 features with the largest variation (i.e., VAR) in the estimated cell-type profiles; (3) DEVarSelect_1VSother for selecting the top 1000 features between one cell type and the other cell types (i.e., 1_O); (4) DEVarSelect_2VSother for selecting the top 1000 features between two cell types and the other cell types (i.e., 2_O); (5) DEVarSelect_pairwise for selecting the top 1000 features between one cell type and another (i.e., P_W); (6) DEVarSelect_RFdecd for selecting the top 1000 features between between one cell type and the other cell types, as well as between two cell types and the other cell types (i.e., RFdecd).

`nMarker` is a numeric of number of cell-type specific markers.  

`InitMarker` is initial cell type specific markers.

`TotalIter` is number of iterations.

# Results of `RFdecd`
We can have estimated cell-type proportion:
```{r}
head(res_RFdecd$estProp)
```

# Session info
```{r sessionInfo, echo=FALSE}
sessionInfo()
```
